platform/source/minimal-printf/mbed_printf_implementation.c

```diff
@@ -247,69 +256,75 @@ static void mbed_minimal_formatted_string_void_pointer(char *buffer, size_t leng
- * @param[in]  value   The value to be printed.
+ * @param      buffer          The buffer to store output (NULL for stdout).
+ * @param[in]  length          The length of the buffer.
+ * @param      result          The current output location.
+ * @param[in]  value           The value to be printed.
+ * @param[in]  dec_precision   The decimal precision. If PRECISION_DEFAULT MBED_CONF_PLATFORM_MINIMAL_PRINTF_SET_FLOATING_POINT_MAX_DECIMALS is used.
+ * @param      width_size       The width modifier.
+ * @param      prepend_zeros    Flag to prepends zeros when the width_size is greater than 0
  */
-static void mbed_minimal_formatted_string_double(char *buffer, size_t length, int *result, double value, FILE *stream)
+static void mbed_minimal_formatted_string_double(char *buffer, size_t length, int *result, double value, int dec_precision, int width_size, bool prepend_zeros, FILE *stream)
 {
     /* get integer part */
     MBED_SIGNED_STORAGE integer = value;
 
-    /* write integer part */
-    mbed_minimal_formatted_string_signed(buffer, length, result, integer, stream);
-
-    /* write decimal point */
-    mbed_minimal_putchar(buffer, length, result, '.', stream);
+    if (dec_precision == PRECISION_DEFAULT) {
+        dec_precision = MBED_CONF_PLATFORM_MINIMAL_PRINTF_SET_FLOATING_POINT_MAX_DECIMALS;
+    }
 
-    /* get decimal part */
-    double precision = 1.0;
+    if (dec_precision != 0) {
+        width_size -= dec_precision + 1; // decimal precision plus '.'
+        if (width_size < 0) {
+            width_size = 0;
+        }
+    } else {
+        value = (value - integer) * 1.0;
+        if (!((value > -0.5) && (value < 0.5))) {
+            integer++;
+        }
+    }
 
-    for (size_t index = 0; index < MBED_CONF_PLATFORM_MINIMAL_PRINTF_SET_FLOATING_POINT_MAX_DECIMALS; index++) {
-        precision *= 10;
+    /* write integer part */
+    if (integer == 0 && value < 0) {
+        mbed_minimal_formatted_string_integer(buffer, length, result, integer, ZERO_NEGATIVE, width_size, prepend_zeros, stream);
+    } else {
+        mbed_minimal_formatted_string_integer(buffer, length, result, integer, INT_SIGNED, width_size, prepend_zeros, stream);
     }
 
-    value = (value - integer) * precision;
+    if (dec_precision != 0) {
+        /* write decimal point */
+        mbed_minimal_putchar(buffer, length, result, '.', stream);
 
-    /* convert to unsigned integer */
-    MBED_UNSIGNED_STORAGE decimal = 0;
+        /* get decimal part */
+        double precision = 1.0;
 
-    if (value < 0) {
-        MBED_SIGNED_STORAGE temp = value;
-        decimal = ~((MBED_UNSIGNED_STORAGE) temp) + 1;
-    } else {
-        decimal = value;
-    }
+        for (size_t index = 0; index < dec_precision; index++) {
+            precision *= 10;
+        }
 
-    /* round up or down */
-    value -= decimal;
+        value = (value - integer) * precision;
 
-    if (!((value > -0.5) && (value < 0.5))) {
-        decimal++;
```

Comment:
The rounding up or down of the decimal part should be kept.

In addition, the original code does not work properly with negative numbers. Given that we have already taken care of the negative sign, we should convert `value` to a positive number. So I'd suggest:

```
        value = (value - integer) * precision;

        /* convert to positive number */
        if (value < 0.0) value *= -1.0;

        MBED_UNSIGNED_STORAGE decimal = value;

        /* round up or down */
        value -= decimal;

        if (!((value > -0.5) && (value < 0.5))) {
            decimal++;
        }

        /* write decimal part */
        mbed_minimal_formatted_string_integer(buffer, length, result, decimal, INT_UNSIGNED, dec_precision, true, stream);
```

Reply:
Done

Reply:
The final code is missing this line:
`value = (value-integer) * precision;`

Reply:
Good catch